### PR TITLE
KIBA-1338: Create api to retrieve collection historical metrics (for graph)

### PR DESCRIPTION
**Pull request by Femi-Ogunkola**

### CODE REVIEW

File: api/notd/model.py [hunk 1]
```diff
@@ -138,3 +138,9 @@ class ProcessedBlock:
     blockHash: str
     blockDate: datetime.datetime
     retrievedTokenTransfers: List[RetrievedTokenTransfer]
+
+@dataclasses.dataclass
+class CollectionGraph:
+    date: datetime.datetime
+    value: int
```
Comment (krishan711) [inline api/notd/model.py:145]: call this tradedValue and tradedAmount

File: api/application.py [hunk 2]
```diff
@@ -1,7 +1,8 @@
 import logging
 import os
 
-from core.api.database_connection_middleware import DatabaseConnectionMiddleware
+from core.api.database_connection_middleware import \
```
Comment (krishan711) [inline api/application.py:4]: single line only pls

File: api/notd/api/api_v1.py [hunk 3]
```diff
@@ -97,6 +99,11 @@ async def get_collection_token_recent_sales(registryAddress: str, tokenId: str,
         tokenTransfers = await notdManager.get_collection_token_recent_sales(registryAddress=registryAddress, tokenId=tokenId, limit=limit, offset=offset)
         return GetCollectionTokenRecentSalesResponse(tokenTransfers=(await responseBuilder.token_transfers_from_models(tokenTransfers=tokenTransfers)))
 
+    @router.get('/collections/{registryAddress}/activity', response_model=GetCollectionActivityResponse)
+    async def get_collection_activity(registryAddress: str):
+        collectionActivities = await notdManager.get_collection_activity(registryAddress=registryAddress)
+        return GetCollectionActivityResponse(collectionActivity=[ApiCollectionActivity(date=collectionActivity.date, value=collectionActivity.tradedValue, amount=collectionActivity.tradedAmount) for collectionActivity in collectionActivities])
```
Comment (krishan711) [inline api/notd/api/api_v1.py:105]: use response builder

File: api/notd/store/retriever.py [hunk 4]
```diff
@@ -177,3 +180,15 @@ async def list_collection_tokens_by_owner(self, address: str, ownerAddress: str,
         for token in soldTokens:
             boughtTokens.remove(token)
         return list(boughtTokens)
+
+    async def get_collection_activity(self, address: str, connection: Optional[DatabaseConnection] = None) -> List[CollectionActivity]:
+        startDate =  date_util.start_of_day(dt=datetime.datetime.now())
+        endDate =  date_util.start_of_day(dt=date_util.datetime_from_datetime(dt=startDate, days=-90))
+        query = select([TokenTransfersTable.c.registryAddress, sqlalchemyfunc.sum(TokenTransfersTable.c.amount), sqlalchemyfunc.sum(TokenTransfersTable.c.value), sqlalchemyfunc.date(BlocksTable.c.blockDate)]).join(BlocksTable, BlocksTable.c.blockNumber == TokenTransfersTable.c.blockNumber)
+        query = query.where(TokenTransfersTable.c.registryAddress == address)
+        query = query.where(BlocksTable.c.blockDate < startDate)
```
Comment (krishan711) [inline api/notd/store/retriever.py:189]: this is wierd - it should already be >= start and < end

File: api/application.py (hunk 2 above)
Comment (Femi-Ogunkola) [reply api/application.py:4]: it was sort --sl that did that, should I still change

File: api/notd/store/retriever.py [hunk 5]
```diff
@@ -177,3 +180,15 @@ async def list_collection_tokens_by_owner(self, address: str, ownerAddress: str,
...
+        query = query.where(BlocksTable.c.blockDate >= endDate)
+        query = query.group_by(sqlalchemyfunc.date(BlocksTable.c.blockDate),TokenTransfersTable.c.registryAddress)
+        result = await self.database.execute(query=query, connection=connection)
+        collectionGraph = [collection_graph_from_row(row) for row in result]
+        return collectionGraph
```
Comment (krishan711) [inline api/notd/store/retriever.py:194]: one thing you should ensure is that data is returned for all dates in the 90 days. if there is nothing in the sql result for that date it should be filled with 0

File: api/application.py (hunk 2 above)
Comment (krishan711) [reply api/application.py:4]: ye, use line length = 1000

File: api/notd/api/api_v1.py [hunk 6]
```diff
@@ -97,6 +98,11 @@ async def get_collection_token_recent_sales(registryAddress: str, tokenId: str,
         tokenTransfers = await notdManager.get_collection_token_recent_sales(registryAddress=registryAddress, tokenId=tokenId, limit=limit, offset=offset)
         return GetCollectionTokenRecentSalesResponse(tokenTransfers=(await responseBuilder.token_transfers_from_models(tokenTransfers=tokenTransfers)))
 
+    @router.get('/collections/{registryAddress}/activity', response_model=GetCollectionActivityResponse)
+    async def get_collection_activity(registryAddress: str):
+        collectionActivities = await notdManager.get_collection_activity(registryAddress=registryAddress)
+        return GetCollectionActivityResponse(collectionActivity= await responseBuilder.collection_activity_from_model(collectionActivities=collectionActivities))
```
Comment (krishan711) [inline api/notd/api/api_v1.py:104]:
```suggestion
        return GetCollectionActivityResponse(collectionActivity=(await responseBuilder.collection_activity_from_model(collectionActivities=collectionActivities)))
```

File: api/notd/api/models_v1.py [hunk 7]
```diff
@@ -60,3 +60,11 @@ class ApiSponsoredToken(BaseModel):
     collection: ApiCollection
     date: datetime.datetime
     latestTransfer: Optional[ApiTokenTransfer]
+
+class ApiCollectionActivity(BaseModel):
+    date: datetime.date
+    totalVolume: int
+    transferCount: int
```
Comment (krishan711) [inline api/notd/api/models_v1.py:67]: these cant be ints in the API - they are too big for JSON int values. they need to be strings

**Pull request closed without merging**